include/stdexec/execution.hpp

```diff
@@ -3236,51 +3285,113 @@ namespace stdexec {
         , __op_state2_(connect(static_cast<_CvrefSender&&>(__sndr), __receiver_t{this})) {
       }
 
-      void __start_op() noexcept {
-        // the inner sender isn't running. if we reach here, then
-        // one way or the other, __shared_state::__notify() will be
-        // called, which decrements the ref count of *this.
-        // So we need to increment it here:
-        this->__inc_ref();
+      void __inc_ref() noexcept {
+        __ref_count_.fetch_add(4, std::memory_order_relaxed);
+      }
+
+      void __dec_ref() noexcept {
+        std::size_t __old_ref_count = __ref_count_.fetch_sub(4, std::memory_order_acq_rel);
+
+        // Here are the old ref counts of significance and what they mean:
+        // 8. There is another waiter that is still using the shared state.
+        // 7. The last operation state has finished using the shared state, and the work has both
+        //    started and completed. The shared state is no longer needed.
+        // 6. Would indicate that the work has completed but was never started. This indicates a
+        //    bug.
+        // 5. The last operation state has finished using the shared state, and the work has not
+        //    completed. The shared state is still needed.
+        // 4. The last waiter has finished using the shared state, and the work has not started and
+        //    never will. The shared state is no longer needed.
+        // 3. Invalid
+        // 2. Invalid
+        // 1. Invalid
+        if (8 > __old_ref_count) {
+          STDEXEC_ASSERT(3 < __old_ref_count && 6 != __old_ref_count);
+
+          // The last waiter has given up the ghost. Detach from the underlying operation.
+          // NOTE: this requests cancellation of the underlying operation.
+          __detach();
```

Comment:
I think this could potentially be a racy access to to the shared-state object.

Consider the case where the underlying operation is started but not yet complete.
We decrement the ref-count and see that this was the last reference.
Before calling `__detach`, the operation completes concurrently on another thread and calls `__notify_waiters`.
This sets the `__completed_bit` then sees there are no more references and thus destroys the shared-state.
The current thread then comes along and calls `__detach()` which now tries to access a destroyed object.

I think to work around this you'll need to set an initial ref-count to 2 (the initial object plus one more).
Then you come in here when the new ref-count reaches 1 (there are no more senders/consumer ops that are referencing it, but the current thread is still referencing it).
The call `__detach()` and then do a final additional ref-count decrement.

Reply:
i'm not sure i'm following where the extra add/dec refs should go

Reply:
Constructor of `__shared_state` should have an initial value for `__ref_count_` of `__ref_count_increment` (or `2 * __ref_count_increment` if you want to avoid the initial increment by  `__data` - see other comment about `__adopt_ref()`).

Then this code would do something like:
```
auto __old_ref_count = __ref_count_.fetch_sub(__ref_count_increment, std::memory_order_acq_rel);
if ((__old_ref_count / __ref_count_increment) == 2) {
  __detach();
  __old_ref_count = __ref_count_.fetch_sub(__ref_count_increment, std::memory_order_acq_rel);
  if ((__old_ref_count & __completed_bit) != 0) {
    assert(__old_ref_count == (__ref_count_increment + __completed_bit + __started_bit));
    delete this;
  }
}
```

i.e. the last call to `__dec_ref()` ends up decrementing the ref-count twice, releasing that initial 1 ref-count, but only after they have called `__detach()`.

Reply:
thanks. i didn't know that "ref-count" meant something other than `__ref_count_`, so i was confuuused.